cpp/src/plasma/external_store.h

```diff
@@ -0,0 +1,124 @@
+// Licensed to the Apache Software Foundation (ASF) under one
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+#ifndef EXTERNAL_STORE_H
+#define EXTERNAL_STORE_H
+
+#include <map>
+#include "plasma/client.h"
+
+namespace plasma {
+
+// ==== The external store ====
+//
+// This file contains declaration for all functions that need to be implemented
+// for an external storage service so that objects evicted from Plasma store
+// can be written to it.
+
+class ExternalStoreHandle {
+ public:
+  ExternalStoreHandle() = default;
+
+  virtual ~ExternalStoreHandle() = default;
+
+  /// This method will be called whenever an object in the Plasma store needs
+  /// to be evicted to the external store.
+  ///
+  /// \param num_objects The number of objects to put.
+  /// \param ids The IDs of the objects to put.
+  /// \param data The object data to put.
+  /// \return The return status.
+  virtual Status Put(size_t num_objects, const ObjectID *ids, const std::string *data) = 0;
+
+  /// This method will be called whenever an evicted object in the External
+  /// store store needs to be accessed.
+  ///
+  /// \param num_objects The number of objects to get.
+  /// \param ids The IDs of the objects to get.
+  /// \param[out] data The object data.
+  /// \return The return status.
+  virtual Status Get(size_t num_objects, const ObjectID *ids, std::string *data) = 0;
+
+};
+
+class ExternalStore {
+ public:
+  /// Default constructor
+  ExternalStore() = default;
+
+  /// Virtual destructor
+  virtual ~ExternalStore() = default;
+
+  /// Connect to the local plasma store. Return the resulting connection.
+  ///
+  /// \param endpoint The name of the endpoint to connect to the external
+  ///        storage service. While the formatting of the endpoint name is
+  ///        specific to the implementation of the external store, it always
+  ///        starts with {store-name}://, where {store-name} is the name of the
+  ///        external store.
+  ///
+  /// \return The return status.
+  virtual std::shared_ptr<ExternalStoreHandle> Connect(const std::string &endpoint) = 0;
```

Comment:
this should again return a Status (to provide further error information if the connect fails. I expect that this could be quite common for external stores that are attached over the network.

```
Status Connect(const std::string &endpoint, std::shared_ptr<ExternalStoreHandle>* store_handle)
```

is a fine signature (where the last parameter is an out parameter).